projects/project2/ImprovAndFutureWork/Algorithm.scala

```diff
@@ -0,0 +1,202 @@
+package com.kg.carl
+
+import com.kg.carl.IdStore._
+import org.apache.spark.rdd.RDD
+import util.control.Breaks._
+import scala.collection.mutable._
+import com.kg.carl.Preprocessor._
+
+object Algorithm {
+
+  case class ScoredRule(p: Int, q: Int, r: Int) {
+    var support = 0
+    var body_support = 0
+    var head_coverage = 0.0
+    var standard_confidence = 0.0
+    var pca_confidence = 0.0
+    var completeness_confidence = 0.0
+    var precision = 0.0
+    var recall = 0.0
+    var directional_metric = 0.0
+    var directional_coef = 0.0
+  }
+
+  val MIN_HEAD_COVERAGE = 0.001
+  val MIN_STANDARD_CONFIDENCE = 0.001
+  val MIN_SUPPORT = 10
+  val CONFIDENCE_INCOMPLETENESS_FACTOR = 0.5
+
+  def doMining(
+    pso:   RDD[(Int, ArrayBuffer[HashMap[Int, ArrayBuffer[Int]]])],
+    nodes: RDD[String], id_for_nodes: RDD[(String, Int)], properties: RDD[Int],
+    expected_cardinalities_by_property_value: RDD[(Int, ArrayBuffer[HashMap[Int, Int]])],
+    numRules:                                 Int): ListBuffer[ScoredRule] = {
+    val entities = Set[Int]()
+    val property_instances_count = LinkedHashMap[Int, Int]().withDefaultValue(0)
+    val entityCount = getNumberOfEntities(nodes)
+    println(entityCount)
+    pso.collect().map {
```

Comment:
Do you really have to collect the `pso`? Or it is necessary for the `property_instances_count` to be updated on the map. Collecting all the values on the driver is considered to be heavy expensive. In case you can map it to another RDD would have been better.